bugbug/comment_features.py

```diff
@@ -0,0 +1,210 @@
+
+                if res is None:
+                    continue
+
+                if isinstance(res, dict):
+                    for key, value in res.items():
+                        data[sys.intern(key)] = value
+                    continue
+
+                if isinstance(res, (list, set)):
+                    for item in res:
+                        data[sys.intern(f"{item} in {feature_extractor_name}")] = True
+                    continue
+
+                data[feature_extractor_name] = res
+
+            commenter_experience_map[comment["creator"]] += 1
+
+            comment_text = comment["text"]
+            for cleanup_function in self.cleanup_functions:
+                comment_text = cleanup_function(comment_text)
+
+            return {
+                "data": data,
+                "comment_text": comment_text,
+            }
+
+        return pd.DataFrame(apply_transform(comment) for comment in comments_iter)
+
+
+class CommenterExperience(CommentFeature):
+    name = "# of Comments made by Commenter in the past"
+
+    def __call__(self, comment, commenter_experience, **kwargs):
+        return commenter_experience
+
+
+class NumberOfLinks(CommentFeature):
+    name = "Number of Links in the comment"
+
+    def __init__(self, domains_to_ignore=set()):
+        self.known_domains = domains_to_ignore
+
+    def __call__(self, comment, **kwargs) -> Any:
+        domains = extract_urls_and_domains(comment["text"])["domains"]
+
+        return {
+            "# of Known links": sum(domain in self.known_domains for domain in domains),
+            "# of Unknown links": sum(
+                domain not in self.known_domains for domain in domains
+            ),
+            "Total # of links": len(domains),
+        }
+
+
+class CharacterCount(CommentFeature):
+    name = "# of Characters in the Comment"
+
+    def __call__(self, comment, **kwargs):
+        return len(comment["text"])
+
+
+class WordCount(CommentFeature):
+    name = "# of Words in the Comment"
+
+    def __call__(self, comment, **kwargs):
+        return len(comment["text"].split())
+
+
+class UnknownLinkAtBeginning(CommentFeature):
+    name = "Unknown Link found at Beginning of the Comment"
+
+    def __init__(self, domains_to_ignore=set()):
+        self.known_domains = domains_to_ignore
+
+    def __call__(self, comment, **kwargs):
+        urls = extract_urls_and_domains(comment["text"], self.known_domains)["urls"]
+
+        words = comment["text"].split()
+        return words[0] in urls if words else False
+
+
+class UnknownLinkAtEnd(CommentFeature):
+    name = "Unknown Link found at End of the Comment"
+
+    def __init__(self, domains_to_ignore=set()):
+        self.known_domains = domains_to_ignore
+
+    def __call__(self, comment, **kwargs):
+        urls = extract_urls_and_domains(comment["text"], self.known_domains)["urls"]
+
+        words = comment["text"].split()
+        return words[-1] in urls if words else False
```

Comment:
Instead, we could return the index for the start of the first link and the index for the end of the last link.

@jpangas wdyt?

Reply:
This sounds good. Let me try it out and I will share the results.